### PR TITLE
Use Apify to parse Instagram URLs

### PR DESCRIPTION
**Pull request by jayjay-w**
## Description

We currently use Apify to parse facebook URLs. In this PR, we are adding the functionality to parse Instagram URLs too using Apify.

References: CV2-4289

## Implementation Plan

 - [x] Update `MediaApifyItem` to accept instagram URLs as well
 - [x] Change the current Instagram parser to call `MediaApifyItem`
 - [x] Update existing tests with expected Apify JSON schema

## How has this been tested?

- Ensured that all instagram URLs for posts and profiles can be parsed
- Existing tests are not broken

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [x] I have added regression tests, if the PR fixes a bug
- [x] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [x] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

### CODE REVIEW

**Review by vasconsaurus — COMMENTED** [review]
Yay! Super excited about this one 🎉 
I tested it locally and it worked, but the embed is the pender one.
So a question: is this expected?
Screenshot 2024-10-31 at 08 59 08

File: app/models/concerns/media_apify_item.rb [hunk 1]
```diff
@@ -5,42 +5,82 @@ class ApifyError < StandardError; end
...
+            { url: url }
+          ]
+        }
+      when :instagram
+        {
+          addParentData: false,
+          directUrls: [url],
+          enhanceUserSearchWithFacebookPage: false,
+          isUserReelFeedURL: false,
+          isUserTaggedFeedURL: false,
```
Comment (caiosba) [inline app/models/concerns/media_apify_item.rb:48]: Question: Can you reliably assume this?

**Review by caiosba — APPROVED** [review]
Tried it locally and worked well, except for the embed issue that @vasconsaurus raised and that we also discussed today, @jayjay-w . I also left one question.

Great work!

**Comment by jayjay-w** [comment]
@vasconsaurus I've added the embed from Instagram, let me know how it looks now:

image

File: app/models/concerns/media_apify_item.rb (hunk 1 above)
Comment (jayjay-w) [reply app/models/concerns/media_apify_item.rb]: Good question. I've removed these checks since we cannot assume this.

**Comment by vasconsaurus** [comment]
@jayjay-w I got mixed results with the embeds:
1. I tried refreshing the media directly in the UI, but it still returned the pender embed (but I think that might be a cache issue and outside of the scope of this ticket)
2. I tried adding new media from two reels posts and a photo post
    1. One reel post returned the instagram embed ([post](https://www.instagram.com/reel/DBePK1SuGDA/)),
    2. the other the pender embed ([post](https://www.instagram.com/reel/DBZKNRquEOJ?utm_source=ig_web_copy_link)), and
    3. the photo post returned `instagram refused to connect` ([post](https://www.instagram.com/p/C-aqRSSu392/?utm_source=ig_web_copy_link))

I checked the responses through swagger and they were all successful.

1. Reel with instagram embed
Screenshot 2024-11-01 at 08 36 28

2. Reel with pender embed
Screenshot 2024-11-01 at 08 36 10

3. Photo post with `instagram refused to connect`
Screenshot 2024-11-01 at 08 47 20

What do you think?

**Comment by vasconsaurus** [comment]
@jayjay-w It mostly worked 🎉 

Worked for:
1. Reels: https://www.instagram.com/reel/DBtq7qVRsKe/?utm_source=ig_web_copy_link
2. Photo: https://www.instagram.com/p/DBjzPG4BIhS/?utm_source=ig_web_copy_link
3. Gallery: https://www.instagram.com/p/DBuFH_lho0U/?utm_source=ig_web_copy_link

But I noticed when I copied the links they had an extra `&igsh=MzRlODBiNWFlZA==`.
I removed for the tests above, but decided to try with it as well, and when it is present, then the pender embed is returned.
1. Photo: https://www.instagram.com/p/DBreZduhJ66/?utm_source=ig_web_copy_link&igsh=MzRlODBiNWFlZA==

**Comment by vasconsaurus** [comment]
@jayjay-w  I tested another one with a similar link to https://www.instagram.com/p/DBreZduhJ66/?utm_source=ig_web_copy_link&igsh=MzRlODBiNWFlZA==, and that one worked 🤔 
https://www.instagram.com/reel/DBjX0GXx67a/?utm_source=ig_web_copy_link&igsh=MzRlODBiNWFlZA==

**Comment by jayjay-w** [comment]
Interesting. We usually remove the `igsh` parameter automatically in the background, for example this is how `https://www.instagram.com/p/DBreZduhJ66/?utm_source=ig_web_copy_link&igsh=MzRlODBiNWFlZA==` looks when I try it out (notice the URL at the bottom does not have `&igsh`, as Pender has stripped it). 

image

**Comment by vasconsaurus** [comment]
@jayjay-w for me it stripped it as well, but returned the pender embed. Since it worked for a similar link, I think it might have been a hiccup on my side 🤔 So, everything looks good to me 🚀

**Review by vasconsaurus — APPROVED** [review]

**Comment by caiosba** [comment]
Great work here! :tada:

**Pull request merged**